File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/stored/library/items/files/job/specs/GivenAQueueOfStoredFileJobs/AndAnUnexpectedTransferErrorOccurs/WhenProcessingTheQueue.java
package com.lasthopesoftware.bluewater.client.stored.library.items.files.job.specs.GivenAQueueOfStoredFileJobs.AndAnUnexpectedTransferErrorOccurs;

import android.os.Build;
import android.support.annotation.RequiresApi;
import com.annimon.stream.Stream;
import com.lasthopesoftware.bluewater.client.library.items.media.files.ServiceFile;
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.StoredFileJob;
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.StoredFileJobProcessor;
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.StoredFileJobState;
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.StoredFileJobStatus;
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.exceptions.StoredFileJobException;
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.specs.GivenAQueueOfStoredFileJobs.MarkedFilesStoredFileAccess;
import com.lasthopesoftware.bluewater.client.stored.library.items.files.repository.StoredFile;
import com.namehillsoftware.handoff.promises.Promise;
import org.junit.BeforeClass;
import org.junit.Test;

import java.io.ByteArrayInputStream;
import java.io.File;
import java.util.*;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class WhenProcessingTheQueue {

	private static final Set<StoredFileJob> storedFileJobs = new HashSet<>(Arrays.asList(
		new StoredFileJob(new ServiceFile(1), new StoredFile().setServiceId(1).setLibraryId(1)),
		new StoredFileJob(new ServiceFile(2), new StoredFile().setServiceId(2).setLibraryId(1)),
		new StoredFileJob(new ServiceFile(4), new StoredFile().setServiceId(4).setLibraryId(1)),
		new StoredFileJob(new ServiceFile(5), new StoredFile().setServiceId(5).setLibraryId(1)),
		new StoredFileJob(new ServiceFile(7), new StoredFile().setServiceId(7).setLibraryId(1)),
		new StoredFileJob(new ServiceFile(114), new StoredFile().setServiceId(114).setLibraryId(1)),
		new StoredFileJob(new ServiceFile(92), new StoredFile().setServiceId(92).setLibraryId(1))));

	private static final StoredFile[] expectedStoredFiles = new StoredFile[] {
		new StoredFile().setServiceId(1).setLibraryId(1),
		new StoredFile().setServiceId(2).setLibraryId(1),
		new StoredFile().setServiceId(4).setLibraryId(1),
	};

	private static final MarkedFilesStoredFileAccess storedFilesAccess = new MarkedFilesStoredFileAccess();

	private static final List<StoredFileJobStatus> storedFileStatuses = new ArrayList<>();
	private static StoredFileJobException exception;

	@RequiresApi(api = Build.VERSION_CODES.N)
	@BeforeClass
	public static void before() {
		final StoredFileJobProcessor storedFileJobProcessor = new StoredFileJobProcessor(
			storedFile -> {
				final File file = mock(File.class);

				when(file.exists()).thenReturn(storedFile.isDownloadComplete());

				if (storedFile.getServiceId() == 5)
					when(file.getPath()).thenReturn("write-failure");

				return file;
			},
			storedFilesAccess,
			f -> new Promise<>(new ByteArrayInputStream(new byte[0])),
			f -> true,
			f -> true,
			(is, f) -> {
				if ("write-failure".equals(f.getPath()))
					throw new UnexpectedException();
			});

			storedFileJobProcessor.observeStoredFileDownload(storedFileJobs).blockingSubscribe(
				storedFileStatuses::add,
				e -> {
					if (e instanceof StoredFileJobException)
						exception = (StoredFileJobException)e;
				});
		}

	@Test
	public void thenTheErrorFileIsCorrect() {
		assertThat(exception.getStoredFile().getServiceId()).isEqualTo(5);
	}

	@Test
	public void thenTheUnexpectedEsceptionIsCorrect() {
		assertThat(exception.getCause()).isInstanceOf(UnexpectedException.class);
	}

	@Test
	public void thenTheFilesAreMarkedAsDownloaded() {
		assertThat(storedFilesAccess.storedFilesMarkedAsDownloaded).containsExactly(expectedStoredFiles);
	}

	@Test
	public void thenTheFilesAreBroadcastAsDownloading() {
		assertThat(Stream.of(storedFileStatuses).filter(s -> s.storedFileJobState == StoredFileJobState.Downloading)
			.map(r -> r.storedFile.getServiceId()).toList())
			.containsExactly(1, 2, 4, 5);
	}

	@Test
	public void thenAllTheFilesAreBroadcastAsDownloaded() {
		assertThat(Stream.of(storedFileStatuses).filter(s -> s.storedFileJobState == StoredFileJobState.Downloaded)
			.map(r -> r.storedFile).toList()).containsExactly(expectedStoredFiles);
	}

	private static class UnexpectedException extends RuntimeException {}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/stored/library/items/files/job/GivenAQueueOfStoredFileJobs/AndTransferringErrorsOut/WhenProcessingTheQueue.kt
package com.lasthopesoftware.bluewater.client.stored.library.items.files.job.GivenAQueueOfStoredFileJobs.AndTransferringErrorsOut

import com.lasthopesoftware.bluewater.client.browsing.files.ServiceFile
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.GivenAQueueOfStoredFileJobs.MarkedFilesStoredFilesUpdater
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.StoredFileJob
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.StoredFileJobProcessor
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.StoredFileJobState
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.StoredFileJobStatus
import com.lasthopesoftware.bluewater.client.stored.library.items.files.repository.StoredFile
import com.lasthopesoftware.promises.extensions.toPromise
import com.lasthopesoftware.resources.io.PromisingReadableStreamWrapper
import com.namehillsoftware.handoff.promises.Promise
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.OutputStream

class WhenProcessingTheQueue {
	private val storedFileJobs = setOf(
		StoredFileJob(
			LibraryId(1),
			ServiceFile("1"),
			StoredFile().setServiceId("1").setLibraryId(1)
		),
		StoredFileJob(
			LibraryId(1),
			ServiceFile("2"),
			StoredFile().setServiceId("2").setLibraryId(1)
		),
		StoredFileJob(
			LibraryId(1),
			ServiceFile("4"),
			StoredFile().setServiceId("4").setLibraryId(1)
		),
		StoredFileJob(
			LibraryId(1),
			ServiceFile("5"),
			StoredFile().setServiceId("5").setLibraryId(1)
		),
		StoredFileJob(
			LibraryId(1),
			ServiceFile("7"),
			StoredFile().setServiceId("7").setLibraryId(1)
		),
		StoredFileJob(
			LibraryId(1),
			ServiceFile("114"),
			StoredFile().setServiceId("114").setLibraryId(1)
		),
		StoredFileJob(
			LibraryId(1),
			ServiceFile("92"),
			StoredFile().setServiceId("92").setLibraryId(1)
		)
	)

	private val expectedStoredFiles = arrayOf(
		StoredFile().setServiceId("1").setLibraryId(1),
		StoredFile().setServiceId("4").setLibraryId(1),
		StoredFile().setServiceId("5").setLibraryId(1),
		StoredFile().setServiceId("7").setLibraryId(1),
		StoredFile().setServiceId("114").setLibraryId(1),
		StoredFile().setServiceId("92").setLibraryId(1)
	)

	private val storedFilesUpdater = MarkedFilesStoredFilesUpdater()
	private var storedFileStatuses: List<StoredFileJobStatus>? = null

	@BeforeAll
	fun before() {
		val storedFileJobProcessor = StoredFileJobProcessor(
			mockk {
				every { promiseOutputStream(any()) } returns ByteArrayOutputStream().toPromise()
				every { promiseOutputStream(match { it.serviceId == "2" }) } returns Promise(mockk<OutputStream>(relaxUnitFun = true) {
					every { write(any(), any(), any()) } throws IOException()
				})
			},
			mockk {
				every { promiseDownload(any(), any()) } answers {
					Promise(PromisingReadableStreamWrapper(ByteArrayInputStream(byteArrayOf(65, 39))))
				}
			},
			storedFilesUpdater,
		)
		storedFileStatuses = storedFileJobProcessor
			.observeStoredFileDownload(storedFileJobs)
			.toList()
			.blockingGet()
	}

	@Test
	fun thenTheErrorFileIsMarkedAsQueued() {
		assertThat(
			storedFileStatuses
				?.filter { s -> s.storedFile.serviceId == "2" }
				?.map { r -> r.storedFileJobState }
		).containsExactly(
			StoredFileJobState.Queued,
			StoredFileJobState.Downloading,
			StoredFileJobState.Queued
		)
	}

	@Test
	fun thenTheFilesAreMarkedAsDownloaded() {
		assertThat(storedFilesUpdater.storedFilesMarkedAsDownloaded).containsExactly(*expectedStoredFiles)
	}

	@Test
	fun thenTheFilesAreBroadcastAsDownloading() {
		assertThat(
			storedFileStatuses
				?.filter { s -> s.storedFileJobState == StoredFileJobState.Downloading }
				?.map { r -> r.storedFile }
		).containsExactlyElementsOf(storedFileJobs.map(StoredFileJob::storedFile))
	}

	@Test
	fun thenAllTheFilesAreBroadcastAsDownloaded() {
		assertThat(
			storedFileStatuses
				?.filter { s -> s.storedFileJobState == StoredFileJobState.Downloaded }
				?.map { r -> r.storedFile }
		).containsExactly(*expectedStoredFiles)
	}
}
